### DISCUSSION TITLE
2 source IPs simultaneously for v4 and v6

### DISCUSSION THREAD

**Discussion by misha4gps** [curl/curl#15038] [Ideas]
In a scenario where I have a v4 and v6 source IPs on different interfaces, I need to use both. Because in general I don't know which target IP will respond v4 or v6, for example:

> Name: google.com
Address: 142.250.74.14
Name: google.com
Address: 2a00:1450:400f:800::200e

Please be so kind as to allow to use both source IPs v4 and v6 simultaneous in a single request

**Comment by bagder** [comment]
By default curl tries both IPv4 and IPv6 when connecting to a server that has both addresses. Those attempts of course have different source addresses.

**Comment by misha4gps** [comment]
Are you thinking about extending the syntax of https://curl.se/libcurl/c/CURLOPT_INTERFACE.html?

**Reply by bagder** [reply]
Are you suggesting that is what you are talking about?

**Reply by misha4gps** [reply]
Not necessarily (it just idea).
I just need to set both source IP addresses to allow receiving data from any URL. Because I don't know what type of destination IP is working now.

**Comment by dimir** [comment]
Let me jump in here. The situation is as follows. Client has many source IPs. When performing an HTTP request it wants to select specific source IPs to be used in the request. And it wants cURL library to do the resolving of the host specified in URL.

Currently it is possible to only specify 1 source IP so the client is limited to using just 1 IP version (whichever the source IP is).

What client wants is to ask cURL library to use whichever source IP suits the IP version of the IP that the host was resolved to.

Given source IP is set to "127.0.0.1,::1"

Scenario 1:
The host was resolved to 127.0.0.2, use source IP 127.0.0.1

Scenario 2:
The host was resolved to ::2, use source IP ::1

So, how hard would it be to teach cURL library to randomly select the source IP from the list respecting the IP version of the resolved one? We know cURL respects IP version when resolving so that shouldn't be too hard I guess.

**Reply by dfandrich** [reply]
If the application needs this level of control, it's always possible for it to perform this job itself. That would mean the application would perform the name resolution irself and provide the exact source and destination addresses to libcurl which would perform the transfer.  That means that it would also need to do its own Happy Eyeballs (or a more targeted approach) if it's possible some hosts won't respond on some of their addresses.

**Reply by dimir** [reply]
OK, I see your point here and it makes sense. Too bad our application currently does not support resolving mechanism and introducing it would not be that easy.
